### PR TITLE
WIP: Memory usage

### LINKED ISSUE #35: reduce memory usage?
Previous versions were working pretty well on a stm32f103, with 20Kb of sram. Using the current version I get segfaults trying to parse even one line of gcode (it works fine on a chip with more memory). I noticed the default settings for the number of arguments in a line of gcode and number of commands in a block were both 10. This takes up about 4Kb per block. Are those sizes arbitrary? or were they based on typical usage? Here's the code I used to test https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9728a5e97e95fd9d27e287b176696e4a . If setting both ArrayVecs down to 5 is ok it drops ram usage to a little over 1Kb.

### PR DESCRIPTION
**Pull request by Michael-F-Bryan**
Refactor the crate to reduce the resource usage for memory-constrained devices.

Fixes #35.

### CODE REVIEW

**Comment by Michael-F-Bryan** [comment]
I just started on the `spans` feature flag for removing the `Span` property from various entities (arguments, gcodes, blocks, etc.) and so far it looks more complicated than just sprinkling a couple `#[cfg(feature = "spans")]` annotations around...

There are lots of places where span information is passed in as part of the constructor, or where we pass around span information while parsing. I'm not a big fan of 20-30 extra `#[cfg]` annotations cluttering up the code base and destroying readability... Do you have any better ideas, @etromby?

**Comment by etrombly** [comment]
For my use case it works with just the tweaks to buffer sizes. I'm not sure if you're interested in supporting devices with less than 20Kb of ram. I think in that case it may make more sense to parse the gcode on a desktop system and pass the movements in a binary format. That should be a simple fix for now. I don't have any experience with writing parsers, so don't really have any input on the spans.

**Comment by Michael-F-Bryan** [comment]
Here are the results of some experiments I've done based on the playground example in the original issue.

**TL;DR:** It looks like limiting the buffer sizes to 5 arguments/command, 1 command/block, and 1 comment/block will be sufficient.

---

([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c99ca1f2960646776b60894114e94eeb))

Original setup (with spans, 10 arguments per command, 10 commands per block, 3 comments per block):

```
Span = 24
Argument = 32
Arguments = 336
Gcode = 384
Commands = 3856
Comment = 40
Comments = 136
Block = 4040
```

Spans commented out/removed (no spans, 10 arguments, 10 commands, 3 comments):

```
Span = 0
Argument = 8
Arguments = 88
Gcode = 112
Commands = 1136
Comment = 16
Comments = 64
Block = 1224
```

Less arguments (with spans, 5 arguments, 10 commands, 3 comments):

```
Span = 24
Argument = 32
Arguments = 176
Gcode = 224
Commands = 2256
Comment = 40
Comments = 136
Block = 2440
```

Decrease all magic numbers (with spans, 5 arguments, 1 command, 1 comment):

```
Span = 24
Argument = 32
Arguments = 176
Gcode = 224
Commands = 240
Comment = 40
Comments = 56
Block = 344
```

No spans **and** decreased magic numbers:

```
Span = 0
Argument = 8
Arguments = 48
Gcode = 72
Commands = 88
Comment = 16
Comments = 32
Block = 144
```

**Comment by etrombly** [comment]
Looks good. If you need to tweak the numbers back up in the future it's an easy change. Thanks

**Comment by Michael-F-Bryan** [comment]
@etrombly I've gone through and tweaked the magic numbers for [commands and comments][1], and [arguments][2]. How do they look to you?

[1]: https://github.com/Michael-F-Bryan/gcode-rs/blob/61ec1150f4391e1d56b3775953ebbccda789f74f/gcode/src/types.rs#L127-L161
[2]: https://github.com/Michael-F-Bryan/gcode-rs/blob/61ec1150f4391e1d56b3775953ebbccda789f74f/gcode/src/types.rs#L258-L278

**Comment by etrombly** [comment]
The documentation for arguments say 1 argument rather than 5. Other than that I think it looks good.

**Comment by Michael-F-Bryan** [comment]
[Released](https://crates.io/crates/gcode) :tada:
